crates/pera-core/src/dwallet_mpc/mpc_manager.rs

```diff
@@ -0,0 +1,665 @@
+        .ok_or_else(|| {
+            PeraError::InvalidCommittee(
+                "received a proof MPC message from a validator that is not in the committee"
+                    .to_string(),
+            )
+        })? as PartyID)
+}
+/// Signature MPC Session Instance.
+/// It keeps track of the session status, the channel to send messages to the instance,
+/// and the messages that are pending to be sent to the instance.
+struct DwalletMPCInstance<Party: CreatableParty> {
+    status: MPCSessionStatus<Party::OutputValue>,
+    /// The messages that are pending to be executed while advancing the instance.
+    /// We need to accumulate a threshold of those before advancing the instance.
+    pending_messages: HashMap<PartyID, Party::Message>,
+    consensus_adapter: Arc<dyn SubmitToConsensus>,
+    epoch_store: Weak<AuthorityPerEpochStore>,
+    epoch_id: EpochId,
+    /// The total number of parties in the chain.
+    /// We can calculate the threshold and party's IDs (indexes) from it.
+    /// To calculate the party's ID, all we need to know is the number of parties,
+    /// as the IDs are just the indexes of those parties.
+    /// If there are three parties, the IDs are [0, 1, 2].
+    number_of_parties: usize,
+    session_id: ObjectID,
+    sender_address: PeraAddress,
+    /// The MPC party being used to run the MPC cryptographic steps.
+    /// `Option` because it can be `None` before the instance has started.
+    party: Option<Party>,
+}
+
+impl<P: CreatableParty> DwalletMPCInstance<P> {
+    fn new(
+        consensus_adapter: Arc<dyn SubmitToConsensus>,
+        epoch_store: Weak<AuthorityPerEpochStore>,
+        epoch: EpochId,
+        session_id: ObjectID,
+        sender_address: PeraAddress,
+        number_of_parties: usize,
+    ) -> Self {
+        Self {
+            status: MPCSessionStatus::Active,
+            pending_messages: HashMap::new(),
+            consensus_adapter: consensus_adapter.clone(),
+            epoch_store: epoch_store.clone(),
+            epoch_id: epoch,
+            session_id,
+            sender_address,
+            party: None,
+            number_of_parties,
+        }
+    }
+
+    fn epoch_store(&self) -> PeraResult<Arc<AuthorityPerEpochStore>> {
+        self.epoch_store
+            .upgrade()
+            .ok_or(PeraError::EpochEnded(self.epoch_id))
+    }
+
+    /// Advances the MPC instance by processing the accumulated messages
+    /// or initializing a new party if needed.
+    ///
+    /// This function handles progressing the MPC session.
+    /// If the session already has an existing
+    /// party instance, it continues the process with it.
+    /// If not, it creates a new party based on the
+    /// participating validators, and their assigned IDs.
+    fn advance(&mut self, auxiliary_input: &P::AuxiliaryInput) -> PeraResult<()> {
+        // Take ownership of the current party, or create a new one if it's the first advance.
```

Comment:
I don't understand why `take()` is needed here. Why not just match on `self.party`?

Reply:
If you try to match directly on self.party (without .take()), Rust will try to borrow self.party temporarily to access its contents. However, because advance() consumes its self parameter, you would need to take full ownership of self.party, which is only possible by moving it out using .take().

Also, you can't match on the party since the match will take a ref and then you will have to `clone()` it in order to use it in advance() and this party can't be cloned + it's even less efficient.

So the options are:

1. change the 2PC-MPC `advance()` api to take a mutable ref (will make the whole code more efficient).
2. take().